go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -276,226 +290,62 @@ func (c *Controller) syncServices(key string) error {
-					// Need to ensure if this vip exists in the worker LBs that we remove it
-					// This can happen if the endpoints originally had host eps but now have cluster only ips
-					if err := deleteNodeVIPs([]string{ip}, svcPort.Protocol, svcPort.Port); err != nil {
-						klog.Errorf("Error deleting VIP %s on per node load balancers, error: %v", vip, err)
-						return err
-					}
-				}
-			}
-			// update the tracker with the VIP
-			c.serviceTracker.updateService(name, namespace, vip, svcPort.Protocol, currentLB)
-			// mark the vip as processed
-			vipsTracked.Delete(virtualIPKey(vip, svcPort.Protocol))
-
-			// Node Port
-			if svcPort.NodePort != 0 {
-				// Ensure we only add correct eps to VIP per node
-				if err := createPerNodePhysicalVIPs(utilnet.IsIPv6String(ip), svcPort.Protocol, svcPort.NodePort,
-					eps, isPolicyLocal); err != nil {
-					c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToUpdateOVNLoadBalancer",
-						"Error trying to update OVN LoadBalancer for Service %s/%s: %v",
-						name, namespace, err)
-					return err
-				}
-				nodeIPs, err := getNodeIPs(utilnet.IsIPv6String(ip))
-				if err != nil {
-					return err
-				}
-				for _, nodeIP := range nodeIPs {
-					vip := util.JoinHostPortInt32(nodeIP, svcPort.NodePort)
-					c.serviceTracker.updateService(name, namespace, vip, svcPort.Protocol, loadbalancer.NodeLoadBalancer)
-					// mark the vip as processed
-					vipsTracked.Delete(virtualIPKey(vip, svcPort.Protocol))
-				}
-			}
-
-			// Services ExternalIPs and LoadBalancer.IngressIPs have the same behavior in OVN
-			// so they are aggregated in a slice and processed together.
-			var externalIPs []string
-			// ExternalIP
-			for _, extIP := range service.Spec.ExternalIPs {
-				// only use the IPs of the same ClusterIP family
-				if utilnet.IsIPv6String(extIP) == utilnet.IsIPv6String(ip) {
-					externalIPs = append(externalIPs, extIP)
-				}
-			}
-			// LoadBalancer
-			for _, ingress := range service.Status.LoadBalancer.Ingress {
-				// only use the IPs of the same ClusterIP family
-				if ingress.IP != "" && utilnet.IsIPv6String(ingress.IP) == utilnet.IsIPv6String(ip) {
-					externalIPs = append(externalIPs, ingress.IP)
-				}
-			}
-
-			// reconcile external IPs
-			if len(externalIPs) > 0 {
-				if err := createPerNodeVIPs(externalIPs, svcPort.Protocol, svcPort.Port, eps, isPolicyLocal); err != nil {
-					klog.Errorf("Error in creating ExternalIP/IngressIP for svc %s, target port: %d - %v\n", name, eps.Port, err)
-				}
-				for _, extIP := range externalIPs {
-					vip := util.JoinHostPortInt32(extIP, svcPort.Port)
-					c.serviceTracker.updateService(name, namespace, vip, svcPort.Protocol, loadbalancer.NodeLoadBalancer)
-					// mark the vip as processed
-					vipsTracked.Delete(virtualIPKey(vip, svcPort.Protocol))
-				}
-			}
+	if !c.repair.legacyLBsDeleted() {
+		if err := deleteServiceFromLegacyLBs(service); err != nil {
```

Comment:
I dont know why the legacy lb cleanup is sprinkled around. IMO this should be in repair and only done once at start up.

Reply:
We can't do it at start up; we have to do it as each service is synced. Otherwise we'll disrupt traffic.

As each service is moved to its own load balance, we have to remove that service's vips from the legacy LBs. Otherwise, we'll have colliding vips (between new and old), and it will be non-deterministic which LB wins. The legacy LBs may have old configuration, so we risk outages by sending traffic to out-of-date pods.

Separately, once all vips are migrated, we can delete the legacy LBs. But that may be some time after startup, so that is a distinct phase of repair.

Reply:
@squeed I don't know if I agree.. 

Building sanity checks into the controller to ensure nothing is stale incrementally does make sense because we don't have to have some complex sync logic, but I think here it's adding quite a bit of complexity overhead to the already complex controller, and I think we could still ensure nothing is stale at startup without much(if any) downtime 

Can't we just have a sync function at startup that Figures out the BATCH of transactions which need to take place to get us into the updated state (no old lbs) 

So we would 
1. Get all services already in the lister and loop 
3. Build the up do date "abstract LBs configs" for each service 
4. Build a single batched nbdb transaction which makes all the new LBs with the correct VIPs AND deletes all of the old LBs in one shot? The only thing that could possible happen is a few milliseconds where the OLD and NEW LBs were there together and in that case the endpoints should still be served so downtime would be minimal right? 

IDK if this is something feasible from the ovsdb POV but I know we're already batching a ton of transactions around the code so I thought I would ask :)

Reply:
At present, this is not feasible. This is because ovn-nbctl doesn't support batching create commands. So, there's no big benefit to batching when we first migrate.

If we're starting in an already-migrated cluster, there won't actually be very many changes - just whatever accumulated during the ovnkube-master restart. So there's not much of a point in batching *those* either.

Most of our existing controllers take this structure, and that's OK unless testing indicates a significant bottleneck.